spanner/cloudexecutor/executor/cloud_executor_util.go

```diff
@@ -0,0 +1,332 @@
+func (s *outcomeSender) finishWithError(err error) error {
+	s.buildOutcome()
+	//s.partialOutcome.Status = &status.Status{Code: int32(gstatus.Code(err)), Message: err.Error()}
+	s.partialOutcome.Status = errToStatus(err)
+	return s.flush()
+}
+
+// appendRow adds another row to buffer. If buffer hits its size limit, the buffered rows will be sent back.
+func (s *outcomeSender) appendRow(row *executorpb.ValueList) error {
+	if !s.hasReadResult && !s.hasQueryResult {
+		return spanner.ToSpannerError(status.Error(codes.InvalidArgument, "either hasReadResult or hasQueryResult should be true"))
+	}
+	if s.rowType == nil {
+		return spanner.ToSpannerError(status.Error(codes.InvalidArgument, "rowType should be set first"))
+	}
+	s.buildOutcome()
+	if s.hasReadResult {
+		s.readResult.Row = append(s.readResult.Row, row)
+		s.rowCount++
+	} else if s.hasQueryResult {
+		s.queryResult.Row = append(s.queryResult.Row, row)
+		s.rowCount++
+	}
+	if s.rowCount >= maxRowsPerBatch {
+		return s.flush()
+	}
+	return nil
+}
+
+// buildOutcome will build the partialOutcome if not exists using relevant variables.
+func (s *outcomeSender) buildOutcome() {
+	if s.partialOutcome != nil {
+		return
+	}
+	s.partialOutcome = &executorpb.SpannerActionOutcome{
+		CommitTime: s.timestamp,
+	}
+	if s.hasReadResult {
+		s.readResult = &executorpb.ReadResult{
+			Table:        s.table,
+			Index:        s.index,
+			RowType:      s.rowType,
+			RequestIndex: s.requestIndex,
+		}
+	} else if s.hasQueryResult {
+		s.queryResult = &executorpb.QueryResult{
+			RowType: s.rowType,
+		}
+	}
+}
+
+// flush sends partialOutcome to stream and clear the internal state
+func (s *outcomeSender) flush() error {
+	if s == nil || s.partialOutcome == nil {
+		log.Println("outcomeSender.flush() is called when there is no partial outcome to send. This is an internal error that should never happen")
+		return spanner.ToSpannerError(status.Error(codes.InvalidArgument, "either outcome sender or partial outcome is nil"))
+	}
+	s.partialOutcome.DmlRowsModified = s.rowsModified
+	if s.hasReadResult {
+		s.partialOutcome.ReadResult = s.readResult
+	} else if s.hasQueryResult {
+		s.partialOutcome.QueryResult = s.queryResult
+	}
+	err := s.sendOutcome(s.partialOutcome)
+	s.partialOutcome = nil
+	s.readResult = nil
+	s.queryResult = nil
+	s.rowCount = 0
+	s.rowsModified = []int64{}
+	return err
+}
+
+// sendOutcome sends the given SpannerActionOutcome.
+func (s *outcomeSender) sendOutcome(outcome *executorpb.SpannerActionOutcome) error {
+	log.Printf("sending result %v actionId %d", outcome, s.actionID)
+	resp := &executorpb.SpannerAsyncActionResponse{
+		ActionId: s.actionID,
+		Outcome:  outcome,
+	}
+	err := s.stream.Send(resp)
+	if err != nil {
+		log.Printf("Failed to send outcome with error: %s", err.Error())
+	} else {
+		log.Printf("Sent result %v actionId %d", outcome, s.actionID)
+	}
+	return err
+}
+
+// errToStatus maps cloud error to Status
+func errToStatus(e error) *spb.Status {
```

Comment:
* Make a sub-directory `utility/mapper` and move this under that. This can be named as `ErrorCodeMapper`
* I'm ok with having `cloud_executor_util.go` but generally I feel its too broad a responsibility and a dumping ground for all kind of utility methods. Wherever possible (if you see an independent responsibility), try to carve out to separate files. That way the code will remain maintainable over time.